vms/platformvm/txs/mempool/mempool.go

```diff
@@ -109,19 +111,32 @@ func New(
 
 	bytesAvailableMetric.Set(maxMempoolSize)
 	return &mempool{
+		toEngine:             toEngine,
 		bytesAvailableMetric: bytesAvailableMetric,
+		numTxs:               numTxs,
+		unissuedTxs:          linkedhashmap.New[ids.ID, *txs.Tx](),
+		consumedUTXOs:        set.NewSet[ids.ID](initialConsumedUTXOsSize),
+		droppedTxIDs:         &cache.LRU[ids.ID, error]{Size: droppedTxIDsCacheSize},
 		bytesAvailable:       maxMempoolSize,
+	}, nil
+}
 
-		unissuedTxs: linkedhashmap.New[ids.ID, *txs.Tx](),
-		numTxs:      numTxs,
+func (m *mempool) Iterate(f func(tx *txs.Tx) bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
 
-		droppedTxIDs:  &cache.LRU[ids.ID, error]{Size: droppedTxIDsCacheSize},
-		consumedUTXOs: set.NewSet[ids.ID](initialConsumedUTXOsSize),
-		toEngine:      toEngine,
-	}, nil
+	itr := m.unissuedTxs.NewIterator()
+	for itr.Next() {
+		if !f(itr.Value()) {
+			return
+		}
+	}
```

Comment:
q: should there be a maximum size here? Or a time bound on how long we can lock for iteration?
I wonder what happens with a clogged mempool. It may be not a big concern for the P-chain, but maybe we should consider some kind of space/time cap?

Reply:
Since we're holding the read lock the mempool can't grow while we're iterating so we should be guaranteed to release this lock. I think we can avoid worrying about this now and revisit if it becomes an issue in the future.

Reply:
I was also going to comment on this block (where excessive iteration prevents `Add` from returning) but agree we can follow-up.

In the HyperSDK, I request "batches" from the iterator to avoid this.